### ISSUE TITLE
Regression in nightly build use of `this` type?

### ISSUE THREAD

**Issue by myitcv** [microsoft/TypeScript#12490]
Easiest explained via the repro steps:

```bash
$ cd `mktemp -d`
$ npm init --force
$ npm install --save-dev typescript@2.2.0-dev.20161122
$ grep -r 'onload:' node_modules/typescript/lib/
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: this, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: this, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: this, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: this, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: this, ev: Event) => any;
....
```

But then

```bash
$ npm install --save-dev typescript@2.2.0-dev.20161124
$ grep -r 'onload:' node_modules/typescript/lib/
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: Document, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: HTMLBodyElement, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: HTMLElement, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: HTMLFrameElement, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: HTMLFrameSetElement, ev: Event) => any;
node_modules/typescript/lib/lib.es6.d.ts:    onload: (this: HTMLIFrameElement, ev: Event) => any;
...
```

This has the obvious impact:

```typescript
let image = new Image();

// with typescript@2.2.0-dev.20161122
image.onload // has type HTMLElement.onload: (this: HTMLImageElement, ev: Event) => any

// with typescript@2.2.0-dev.20161124
image.onload // has type HTMLElement.onload: (this: HTMLElement, ev: Event) => any
```

Is this an intentional change?

**Comment by myitcv** [comment]
@mhegazy et al - any thoughts on this?

**Comment by myitcv** [comment]
@mhegazy I'd like to be able to get back to using the latest nightly, help to flush out issues on your side and ours. But because of this issue, the nightly build is broken from our perspective. 

Do you have any indication on when this will be fixed? 

Or are there bigger question that arise out of https://github.com/Microsoft/TypeScript/pull/12433 that need addressing/thinking about first?

Thanks

**Comment by mhegazy** [comment]
i need to update the script that generates the file in https://github.com/Microsoft/TSJS-lib-generator. I have not had a chance to do that. PRs are always welcomed.

**Comment by myitcv** [comment]
@mhegazy this bug is present in the now released 2.1.4 and 2.1.5. 

One thing I'm not clear on is why this isn't breaking for more people?

**Comment by Andarist** [comment]
It's hard to tell if this was fixed or not - the DOM lib types have changed and `this` is typed as `GlobalEventHandlers` within `onload`'s signature. So it seems that if one wants to access a more narrow type for `this` they should use `addEventListener` (and that works OK):
```ts
let image = new Image();

image.addEventListener('load', function () {
  this // HTMLImageElement
})
```

I think that the issue can be closed for housekeeping purposes, cc @RyanCavanaugh
